src/components/sticky-object.js

```diff
@@ -0,0 +1,123 @@
+/* global THREE, CANNON, AFRAME */
+AFRAME.registerComponent("sticky-object", {
+  dependencies: ["body", "super-networked-interactable"],
+
+  schema: {
+    autoLockOnLoad: { default: false },
+    autoLockOnRelease: { default: false },
+    autoLockSpeedLimit: { default: 0.25 }
+  },
+
+  init() {
+    this._onGrab = this._onGrab.bind(this);
+    this._onRelease = this._onRelease.bind(this);
+    this._onBodyLoaded = this._onBodyLoaded.bind(this);
+  },
+
+  play() {
+    this.el.addEventListener("grab-start", this._onGrab);
+    this.el.addEventListener("grab-end", this._onRelease);
+    this.el.addEventListener("body-loaded", this._onBodyLoaded);
+  },
+
+  pause() {
+    this.el.removeEventListener("grab-start", this._onGrab);
+    this.el.removeEventListener("grab-end", this._onRelease);
+    this.el.removeEventListener("body-loaded", this._onBodyLoaded);
+  },
+
+  setLocked(locked) {
+    if (!NAF.utils.isMine(this.el)) return;
+
+    const mass = this.el.components["super-networked-interactable"].data.mass;
+    this.locked = locked;
+    this.el.body.type = locked ? window.CANNON.Body.STATIC : window.CANNON.Body.DYNAMIC;
+    this.el.setAttribute("body", {
+      mass: locked ? 0 : mass
+    });
+  },
+
+  _onBodyLoaded() {
+    if (this.data.autoLockOnLoad) {
+      this.setLocked(true);
+    }
+  },
+
+  _onRelease() {
+    if (
+      this.data.autoLockOnRelease &&
+      this.el.body.velocity.lengthSquared() < this.data.autoLockSpeedLimit * this.data.autoLockSpeedLimit
+    ) {
+      this.setLocked(true);
+    }
+  },
+
+  _onGrab() {
+    this.setLocked(false);
+  },
+
+  remove() {
+    if (this.stuckTo) {
+      const stuckTo = this.stuckTo;
+      delete this.stuckTo;
+      stuckTo._unstickObject();
+    }
+  }
+});
+
+AFRAME.registerComponent("sticky-object-zone", {
+  dependencies: ["physics"],
+  init() {
+    // TODO: position/rotation/impulse need to get updated if the sticky-object-zone moves
+    this.worldQuaternion = new THREE.Quaternion();
+    this.worldPosition = new THREE.Vector3();
+    this.el.object3D.getWorldQuaternion(this.worldQuaternion);
+    this.el.object3D.getWorldPosition(this.worldPosition);
+
+    const dir = new THREE.Vector3(0, 0, 5).applyQuaternion(this.el.object3D.quaternion);
+    this.bootImpulsePosition = new CANNON.Vec3(0, 0, 0);
+    this.bootImpulse = new CANNON.Vec3();
+    this.bootImpulse.copy(dir);
+
+    this.el.addEventListener("collisions", e => {
```

Comment:
This listener should be cleaned up in `remove` (move the callback into its own bound function?)